haematite_ser/src/wrap.rs

```diff
@@ -0,0 +1,242 @@
+use std::collections::HashMap;
+
+use haematite_models::meta::permissions::Tree;
+use serde::ser::{Serialize, Serializer};
+use serde::ser::{
+    SerializeMap, SerializeSeq, SerializeStruct, SerializeStructVariant, SerializeTuple,
+    SerializeTupleStruct, SerializeTupleVariant,
+};
+
+#[derive(Debug, Clone)]
+pub struct SerializeWrap {
+    pub inner: WrapType,
+    pub allowed: bool,
+}
+
+impl SerializeWrap {
+    pub fn new(inner: WrapType) -> Self {
+        Self {
+            inner,
+            allowed: true,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum WrapType {
+    Bool(bool),
+    I8(i8),
+    I16(i16),
+    I32(i32),
+    I64(i64),
+    U8(u8),
+    U16(u16),
+    U32(u32),
+    U64(u64),
+    F32(f32),
+    F64(f64),
+    Char(char),
+    Str(String),
+    Bytes(Vec<u8>),
+    None,
+    Some(Box<WrapType>),
+    Unit,
+    UnitStruct(&'static str),
+    UnitVariant(&'static str, u32, &'static str),
+    NewtypeStruct(&'static str, Box<WrapType>),
+    NewtypeVariant(&'static str, u32, &'static str, Box<WrapType>),
+    Map(HashMap<String, SerializeWrap>),
+    Seq(Vec<SerializeWrap>),
+    Struct(&'static str, HashMap<&'static str, SerializeWrap>),
+    StructVariant(
+        &'static str,
+        u32,
+        &'static str,
+        HashMap<&'static str, SerializeWrap>,
+    ),
+    Tuple(Vec<SerializeWrap>),
+    TupleStruct(&'static str, Vec<SerializeWrap>),
+    TupleVariant(&'static str, u32, &'static str, Vec<SerializeWrap>),
+}
+
+impl WrapType {
+    pub fn update_with(&mut self, tree: &Tree) {
+        if let Tree::InternalVertex(perm_map) = tree {
+            match self {
+                Self::Map(map) => {
+                    for (key, value) in map.iter_mut() {
+                        if let Some(subtree) = perm_map.get(key) {
+                            value.allowed = true;
+                            value.inner.update_with(subtree);
+                        } else {
+                            value.allowed = false;
+                        }
+                    }
+                }
+                Self::Seq(values) => {
+                    for (i, value) in values.iter_mut().enumerate() {
+                        if let Some(subtree) = perm_map.get(&i.to_string()) {
```

Comment:
I would expect a sequence (as opposed to a tuple) to be homogeneous and of unpredictable length, so separate per-element permissions don't seem to make sense.